File: PYTHON/Pertemuan 02/ElemenDasarJava.py
import sys
sys.stdout.reconfigure(encoding='utf-8')
import math
from Mahasiswa import Mahasiswa

# ── Konstanta program ────────────────────────────────────────────────────
PI = 3.14159265
MAX_NILAI = 100

# ── Demo Tipe Data Primitif ──────────────────────────────────────────────
def demo_primitive():
    print("\n─── TIPE DATA PRIMITIF ───────────────────────")
    b = 127            # Python only has int and float, no byte/short
    s = 32_000
    i = 2_000_000
    l = 9_000_000_000
    f = 3.14
    d = 3.14159265
    ok = True
    c = 'A'

    print(f"  byte   b  = {b}")
    print(f"  short  s  = {s}")
    print(f"  int    i  = {i}")
    print(f"  long   l  = {l}")
    print(f"  float  f  = {f}")
    print(f"  double d  = {d}")
    print(f"  boolean ok= {ok}")
    print(f"  char   c  = {c}  (kode ASCII: {ord(c)})")

# ── Demo Operator ────────────────────────────────────────────────────────
def demo_operator():
    print("\n─── OPERATOR ────────────────────────────────")
    a, b = 10, 3

    print("Aritmatika (a=10, b=3):")
    print(f"  a+b={a+b}  a-b={a-b}  a*b={a*b}")
    print(f"  a/b={a//b} (integer)  (float)a/b={a/b}")
    print(f"  a%b={a%b} (sisa bagi)")

    print("Relasional:")
    print(f"  a>b:{a>b}  a<b:{a<b}  a==b:{a==b}  a!=b:{a!=b}")

    print("Logika:")
    print(f"  true && false = {True and False}")
    print(f"  true || false = {True or False}")
    print(f"  !true = {not True}")

    print("Assignment singkat:")
    x = 10
    x += 5
    print(f"  x=10, x+=5 → x={x}")
    x -= 3
    print(f"  x-=3 → x={x}")
    x *= 2
    print(f"  x*=2 → x={x}")

# ── Demo Kontrol Alur ────────────────────────────────────────────────────
def demo_kontrol_alur():
    print("\n─── KONTROL ALUR ────────────────────────────")

    # if-else: hitung grade
    nilai_list = [92, 78, 65, 50, 40]
    print("if-else – Grade Mahasiswa:")
    for nilai in nilai_list:
        if nilai >= 85:
            grade = "A"
        elif nilai >= 70:
            grade = "B"
        elif nilai >= 55:
            grade = "C"
        elif nilai >= 40:
            grade = "D"
        else:
            grade = "E (Tidak Lulus)"
        print(f"  Nilai {nilai:.0f} → Grade {grade}")

    # match (Python 3.10+) equivalent to switch
    print("switch (match) – Nama Hari:")
    for hari in range(1, 8):
        match hari:
            case 1: print("Senin ", end="")
            case 2: print("Selasa ", end="")
            case 3: print("Rabu ", end="")
            case 4: print("Kamis ", end="")
            case 5: print("Jumat ", end="")
            case 6: print("Sabtu ", end="")
            case 7: print("Minggu ", end="")
    print()

    # for – menghitung FPB
    p, q = 48, 18
    print(f"for – FPB({p},{q}):")
    fpb = 1
    for k in range(1, min(p, q) + 1):
        if p % k == 0 and q % k == 0:
            fpb = k
    print(f"  FPB = {fpb}")

    # while – bilangan prima
    print("while – Bilangan prima ≤ 30: ", end="")
    num = 2
    while num <= 30:
        prima = True
        for k in range(2, int(math.sqrt(num)) + 1):
            if num % k == 0:
                prima = False
                break
        if prima:
            print(f"{num} ", end="")
        num += 1
    print()

    # do-while (simulated in Python)
    print("do-while – Pangkat 2 < 1000: ", end="")
    pot = 1
    while True:
        print(f"{pot} ", end="")
        pot *= 2
        if pot >= 1000:
            break
    print()

# ── Demo Coding Standard ─────────────────────────────────────────────────
def demo_coding_standard():
    print("\n─── PYTHON CODING STANDARD (PEP 8) ────────────────────")
    print("  Class      PascalCase  : MahasiswaBaru, RekeningTabungan")
    print("  Function   snake_case  : hitung_ipk(), get_nama(), set_umur()")
    print("  Variabel   snake_case  : nama_lengkap, total_sks, ipk_akhir")
    print("  Konstanta  UPPER_SNAKE : MAX_NILAI, PI, MIN_SKS_LULUS")
    print("  Package    lowercase   : com.uph.pbo, id.ac.uph.si")
    print()
    print("  ✗ Salah : class mahasiswa_baru, int Nilai_Akhir")
    print("  ✓ Benar : class MahasiswaBaru, nilai_akhir = 10")

# ── Demo Access Modifier ─────────────────────────────────────────────────
def demo_access_modifier():
    print("\n─── ACCESS MODIFIER (Konvensi Python) ─────────────────────────")
    print("Python tidak memiliki access modifier sejati. Menggunakan konvensi:")
    print("  - public: nama_variabel")
    print("  - protected: _nama_variabel (hanya konvensi untuk subclass)")
    print("  - private: __nama_variabel (name mangling)")

# ── main ─────────────────────────────────────────────────────────────────
def main():
    print("╔══════════════════════════════════════════════╗")
    print("║   PERTEMUAN 2 – Elemen Dasar PBO dengan Python║")
    print("╚══════════════════════════════════════════════╝")

    demo_primitive()
    demo_operator()
    demo_kontrol_alur()
    demo_coding_standard()
    demo_access_modifier()

    # Demo class Mahasiswa
    print("\n─── DEMO CLASS Mahasiswa ───────────")
    mhs_list = [
        Mahasiswa("Budi", 21, 3.50),
        Mahasiswa("Sari", 20, 1.80),
        Mahasiswa("Andi", 22, 2.75)
    ]
    for m in mhs_list:
        m.tampilkan()

    print(f"\n[Konstanta: PI={PI}, MAX_NILAI={MAX_NILAI}]")

if __name__ == "__main__":
    main()

File: JAVA/Pertemuan 02/src/id/ac/uph/pbo/pertemuan02/ElemenDasarJava.java
package id.ac.uph.pbo.pertemuan02;

public class ElemenDasarJava {

    // ── Konstanta program ────────────────────────────────────────────────────
    static final double PI        = 3.14159265;
    static final int    MAX_NILAI = 100;

    // ── Demo Tipe Data Primitif ──────────────────────────────────────────────
    static void demoPrimitive() {
        System.out.println("\n─── TIPE DATA PRIMITIF ───────────────────────");
        byte   b  = 127;            // 8-bit  : -128 s/d 127
        short  s  = 32_000;         // 16-bit : -32768 s/d 32767
        int    i  = 2_000_000;      // 32-bit
        long   l  = 9_000_000_000L; // 64-bit (huruf L di akhir)
        float  f  = 3.14f;          // 32-bit desimal (huruf f)
        double d  = 3.14159265;     // 64-bit desimal (lebih presisi)
        boolean ok = true;
        char   c  = 'A';

        System.out.println("  byte   b  = " + b);
        System.out.println("  short  s  = " + s);
        System.out.println("  int    i  = " + i);
        System.out.println("  long   l  = " + l);
        System.out.println("  float  f  = " + f);
        System.out.println("  double d  = " + d);
        System.out.println("  boolean ok= " + ok);
        System.out.println("  char   c  = " + c + "  (kode ASCII: " + (int)c + ")");
    }

    // ── Demo Operator ────────────────────────────────────────────────────────
    static void demoOperator() {
        System.out.println("\n─── OPERATOR ────────────────────────────────");
        int a = 10, b = 3;

        System.out.println("Aritmatika (a=10, b=3):");
        System.out.println("  a+b=" + (a+b) + "  a-b=" + (a-b) + "  a*b=" + (a*b));
        System.out.println("  a/b=" + (a/b) + " (integer)  (double)a/b=" + ((double)a/b));
        System.out.println("  a%b=" + (a%b) + " (sisa bagi)");

        System.out.println("Relasional:");
        System.out.println("  a>b:" + (a>b) + "  a<b:" + (a<b) + "  a==b:" + (a==b) + "  a!=b:" + (a!=b));

        System.out.println("Logika:");
        System.out.println("  true && false = " + (true && false));
        System.out.println("  true || false = " + (true || false));
        System.out.println("  !true = " + (!true));

        System.out.println("Assignment singkat:");
        int x = 10;
        x += 5; System.out.println("  x=10, x+=5 → x=" + x);
        x -= 3; System.out.println("  x-=3 → x=" + x);
        x *= 2; System.out.println("  x*=2 → x=" + x);
    }

    // ── Demo Kontrol Alur ────────────────────────────────────────────────────
    static void demoKontrolAlur() {
        System.out.println("\n─── KONTROL ALUR ────────────────────────────");

        // if-else: hitung grade
        double[] nilaiList = {92, 78, 65, 50, 40};
        System.out.println("if-else – Grade Mahasiswa:");
        for (double nilai : nilaiList) {
            String grade;
            if      (nilai >= 85) grade = "A";
            else if (nilai >= 70) grade = "B";
            else if (nilai >= 55) grade = "C";
            else if (nilai >= 40) grade = "D";
            else                  grade = "E (Tidak Lulus)";
            System.out.printf("  Nilai %.0f → Grade %s%n", nilai, grade);
        }

        // switch
        System.out.println("switch – Nama Hari:");
        for (int hari = 1; hari <= 7; hari++) {
            switch (hari) {
                case 1: System.out.print("Senin "); break;
                case 2: System.out.print("Selasa "); break;
                case 3: System.out.print("Rabu "); break;
                case 4: System.out.print("Kamis "); break;
                case 5: System.out.print("Jumat "); break;
                case 6: System.out.print("Sabtu "); break;
                case 7: System.out.print("Minggu "); break;
            }
        }
        System.out.println();

        // for – menghitung FPB
        int p = 48, q = 18;
        System.out.println("for – FPB(" + p + "," + q + "):");
        int fpb = 1;
        for (int k = 1; k <= Math.min(p, q); k++) {
            if (p % k == 0 && q % k == 0) fpb = k;
        }
        System.out.println("  FPB = " + fpb);

        // while – bilangan prima
        System.out.print("while – Bilangan prima ≤ 30: ");
        int num = 2;
        while (num <= 30) {
            boolean prima = true;
            for (int k = 2; k <= Math.sqrt(num); k++) {
                if (num % k == 0) { prima = false; break; }
            }
            if (prima) System.out.print(num + " ");
            num++;
        }
        System.out.println();

        // do-while
        System.out.print("do-while – Pangkat 2 < 1000: ");
        int pot = 1;
        do {
            System.out.print(pot + " ");
            pot *= 2;
        } while (pot < 1000);
        System.out.println();
    }

    // ── Demo Coding Standard ─────────────────────────────────────────────────
    static void demoCodingStandard() {
        System.out.println("\n─── JAVA CODING STANDARD ────────────────────");
        System.out.println("  Class      PascalCase  : MahasiswaBaru, RekeningTabungan");
        System.out.println("  Method     camelCase   : hitungIPK(), getNama(), setUmur()");
        System.out.println("  Variabel   camelCase   : namaLengkap, totalSKS, ipkAkhir");
        System.out.println("  Konstanta  UPPER_SNAKE : MAX_NILAI, PI, MIN_SKS_LULUS");
        System.out.println("  Package    lowercase   : com.uph.pbo, id.ac.uph.si");
        System.out.println();
        System.out.println("  ✗ Salah : class mahasiswa_baru, int Nilai_Akhir");
        System.out.println("  ✓ Benar : class MahasiswaBaru, int nilaiAkhir");
    }

    // ── Demo Access Modifier ─────────────────────────────────────────────────
    static void demoAccessModifier() {
        System.out.println("\n─── ACCESS MODIFIER ─────────────────────────");
        System.out.printf("  %-12s %-12s %-12s %-12s %-12s%n",
                "Modifier", "Class ini", "Package", "Subclass", "Semua");
        System.out.println("  " + "─".repeat(60));
        System.out.printf("  %-12s %-12s %-12s %-12s %-12s%n",
                "public",    "✓", "✓", "✓", "✓");
        System.out.printf("  %-12s %-12s %-12s %-12s %-12s%n",
                "protected", "✓", "✓", "✓", "✗");
        System.out.printf("  %-12s %-12s %-12s %-12s %-12s%n",
                "(default)", "✓", "✓", "✗", "✗");
        System.out.printf("  %-12s %-12s %-12s %-12s %-12s%n",
                "private",   "✓", "✗", "✗", "✗");
    }

    // ── main ─────────────────────────────────────────────────────────────────
    public static void main(String[] args) {
        System.out.println("╔══════════════════════════════════════════════╗");
        System.out.println("║   PERTEMUAN 2 – Elemen Dasar PBO dengan Java  ║");
        System.out.println("╚══════════════════════════════════════════════╝");

        demoPrimitive();
        demoOperator();
        demoKontrolAlur();
        demoCodingStandard();
        demoAccessModifier();

        // Demo class Mahasiswa
        System.out.println("\n─── DEMO CLASS Mahasiswa ───────────");
        Mahasiswa[] mhsList = {
            new Mahasiswa("Budi",  21, 3.50),
            new Mahasiswa("Sari",  20, 1.80),
            new Mahasiswa("Andi",  22, 2.75)
        };
        for (Mahasiswa m : mhsList) m.tampilkan();

        System.out.println("\n[Konstanta: PI=" + PI + ", MAX_NILAI=" + MAX_NILAI + "]");
    }
}
